File: api/trigger.py
import json
import urllib.request
import urllib.parse

GHL_PIT      = "pit-d977607c-f9e5-4a92-8b48-07d2ea342d79"
GHL_LOCATION = "KDMRygh4EQxW5HYf5SZu"
WF_CALL      = "5a799e8b-34c1-483d-8999-06fa8d3fe0ae"
WF_SMS       = "1803b643-fc48-4f8e-a98a-a7728878320e"

def ghl_req(method, path, body=None, params=None):
    url = "https://services.leadconnectorhq.com" + path
    if params:
        url += "?" + urllib.parse.urlencode(params)
    data = json.dumps(body).encode() if body else None
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header("Authorization", f"Bearer {GHL_PIT}")
    req.add_header("Version", "2023-02-21")
    req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req, timeout=10) as r:
            return r.status, json.loads(r.read())
    except urllib.error.HTTPError as e:
        return e.code, json.loads(e.read())

def get_or_create_contact(first_name, last_name, phone):
    status, data = ghl_req("GET", "/contacts/search/duplicate",
        params={"locationId": GHL_LOCATION, "number": phone})
    if status == 200:
        contacts = data.get("contacts", [])
        if contacts:
            return contacts[0]["id"], False
    status, data = ghl_req("POST", "/contacts/", {
        "locationId": GHL_LOCATION, "firstName": first_name,
        "lastName": last_name, "phone": phone
    })
    if status in (200, 201):
        cid = data.get("contact", {}).get("id") or data.get("id")
        return cid, True
    if status == 400:
        cid = data.get("meta", {}).get("contactId")
        if cid:
            return cid, False
    return None, False

def add_to_workflow(contact_id, workflow_id):
    status, _ = ghl_req("POST", f"/contacts/{contact_id}/workflow/{workflow_id}",
        {"eventStartTime": None})
    return status in (200, 201, 202), status

def handler(request):
    if request.method == "OPTIONS":
        return Response("", status=200, headers={
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Methods": "POST",
            "Access-Control-Allow-Headers": "Content-Type"
        })
    body = json.loads(request.body)
    first_name = body.get("first_name", "")
    last_name  = body.get("last_name", "")
    phone      = body.get("phone", "")
    action     = body.get("action", "both")
    steps = []

    contact_id, created = get_or_create_contact(first_name, last_name, phone)
    if contact_id:
        steps.append({"ok": True, "msg": f"Contact {'created' if created else 'found'}: {contact_id}"})
    else:
        steps.append({"ok": False, "msg": "Failed to create/find contact"})
        return Response(json.dumps({"steps": steps}), status=200,
            headers={"Content-Type": "application/json", "Access-Control-Allow-Origin": "*"})

    if action in ("call", "both"):
        ok, status = add_to_workflow(contact_id, WF_CALL)
        steps.append({"ok": ok, "msg": f"Call workflow: {'OK' if ok else 'FAILED'} ({status})"})

    if action in ("sms", "both"):
        ok, status = add_to_workflow(contact_id, WF_SMS)
        steps.append({"ok": ok, "msg": f"SMS workflow: {'OK' if ok else 'FAILED'} ({status})"})

    return Response(json.dumps({"steps": steps}), status=200,
        headers={"Content-Type": "application/json", "Access-Control-Allow-Origin": "*"})

File: api/trigger.js
const https = require('https');

const GHL_PIT      = 'pit-d977607c-f9e5-4a92-8b48-07d2ea342d79';
const GHL_LOCATION = 'KDMRygh4EQxW5HYf5SZu';
const WF_CALL      = '5a799e8b-34c1-483d-8999-06fa8d3fe0ae';
const WF_SMS       = '1803b643-fc48-4f8e-a98a-a7728878320e';

function ghlReq(method, path, body, params) {
  return new Promise((resolve) => {
    let url = 'services.leadconnectorhq.com';
    let fullPath = path;
    if (params) fullPath += '?' + new URLSearchParams(params).toString();
    const data = body ? JSON.stringify(body) : null;
    const opts = {
      hostname: url, path: fullPath, method,
      headers: {
        'Authorization': `Bearer ${GHL_PIT}`,
        'Version': '2023-02-21',
        'Content-Type': 'application/json',
        ...(data ? {'Content-Length': Buffer.byteLength(data)} : {})
      }
    };
    const req = https.request(opts, (res) => {
      let d = '';
      res.on('data', c => d += c);
      res.on('end', () => { try { resolve({status: res.statusCode, body: JSON.parse(d)}); } catch { resolve({status: res.statusCode, body: {}}); }});
    });
    req.on('error', () => resolve({status: 500, body: {}}));
    if (data) req.write(data);
    req.end();
  });
}

async function getOrCreate(firstName, lastName, phone) {
  const s = await ghlReq('GET', '/contacts/search/duplicate', null, {locationId: GHL_LOCATION, number: phone});
  if (s.status === 200 && s.body.contacts?.length) return [s.body.contacts[0].id, false];
  const c = await ghlReq('POST', '/contacts/', {locationId: GHL_LOCATION, firstName, lastName, phone});
  if ([200,201].includes(c.status)) return [c.body.contact?.id || c.body.id, true];
  if (c.status === 400 && c.body.meta?.contactId) return [c.body.meta.contactId, false];
  return [null, false];
}

async function addToWf(contactId, wfId) {
  const r = await ghlReq('POST', `/contacts/${contactId}/workflow/${wfId}`, {eventStartTime: null});
  return [r.status >= 200 && r.status < 300, r.status];
}

module.exports = async (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  if (req.method === 'OPTIONS') return res.status(200).end();

  const {first_name, last_name, phone, action} = req.body;
  const steps = [];

  const [contactId, created] = await getOrCreate(first_name, last_name || '', phone);
  if (!contactId) {
    steps.push({ok: false, msg: 'Failed to create/find contact'});
    return res.json({steps});
  }
  steps.push({ok: true, msg: `Contact ${created ? 'created' : 'found'}: ${contactId}`});

  if (['call','both'].includes(action)) {
    const [ok, st] = await addToWf(contactId, WF_CALL);
    steps.push({ok, msg: `Call workflow: ${ok ? 'OK' : 'FAILED'} (${st})`});
  }
  if (['sms','both'].includes(action)) {
    const [ok, st] = await addToWf(contactId, WF_SMS);
    steps.push({ok, msg: `SMS workflow: ${ok ? 'OK' : 'FAILED'} (${st})`});
  }
  res.json({steps});
};
